sql/core/src/test/scala/org/apache/spark/sql/ProductAggSuite.scala

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.expressions.Window
+import org.apache.spark.sql.functions.{ col, lit, product }
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.{ByteType, DoubleType, FloatType, IntegerType, ShortType}
+
+
+class ProductAggSuite extends QueryTest with SharedSparkSession {
+  // Sequence of integers small enough that factorial is representable exactly as DoubleType:
+  private lazy val data16 = spark.range(1, 17).toDF("x")
+
+  private lazy val factorials = (1 to 16).scanLeft(1L) { case (f, x) => f * x }
+
+  test("bare factorial") {
+    implicit val enc = Encoders.scalaDouble
+
+    val prod = data16.agg(product(col("x"))).as[Double].head
+    val expected = (1L to 16L).reduce { _ * _ }.toDouble
+
+    assert(prod === expected)
+    assert(prod === factorials(16))
+  }
+
+  test("type flexibility") {
+    val bytes16 = spark.createDataset((1 to 16).map { _.toByte })(Encoders.scalaByte).toDF("x")
```

Comment:
Can we move the tests to `DataFrameAggregateSuite` and `DataFrameWindowFunctionsSuite`, and use the test infra there? Then you don't need to specify the encoder manually and can use `checkAnswer` to check result.

Reply:
Is this a high priority change - it's not clear whether this will make the tests substantially more effective?

Reply:
The test infra will do more checks for you automatically. e.g. it checks the structural integrity of the plan, it checks RDD serialization, etc.

I'm OK if you insist to create a new suite for product, but please follow `DataFrameAggregateSuite` to write tests.

Reply:
Thanks, @cloud-fan - that's useful to know.
For the current (low risk?) `product` aggregation, I'm doubtful that making these changes is justifiable at this stage of development.
Obviously, for any future patches I develop, I'll be happy to build these around `DataFrameAggregateSuite` where appropriate.

Reply:
It's not only about risk, but also about having a consistent code style for writing tests in Spark. If another people is trying to add a new agg function as you did, and sees this test suite,he/she may follow it to write a new test suite, which is unexpected.

If you don't have time to do this, please create a JIRA ticket for rewriting this test suite, add a TODO with the JIRA ID here, so that other people won't follow it to write tests.

Reply:
Ok, I've rebuilt `sql/ProductAggSuite.scala` to be more like `DataFrameAggregateSuite`.